File: src/kitaru/client/resources/evaluators.py
"""Evaluators resource."""

import uuid
from collections.abc import AsyncIterator
from typing import TYPE_CHECKING

from kitaru.api_models.v1.base import ListParams, Page
from kitaru.api_models.v1.evaluator import (
    EvaluatorCreateRequest,
    EvaluatorListParams,
    EvaluatorResponse,
    EvaluatorUpdateRequest,
    EvaluatorVersionCreateRequest,
    EvaluatorVersionResponse,
    EvaluatorVersionUpdateRequest,
)
from kitaru.client.resources.pagination import iterate_pages

if TYPE_CHECKING:
    from kitaru.client.api_client import KitaruAPIClient


class EvaluatorsResource:
    """Evaluator API methods."""

    def __init__(self, client: "KitaruAPIClient") -> None:
        self._client = client

    async def create(self, request: EvaluatorCreateRequest) -> EvaluatorResponse:
        response = await self._client.request(
            "POST",
            "/v1/evaluators",
            json=request.model_dump(mode="json", exclude_unset=True),
        )
        return EvaluatorResponse.model_validate(response.json())

    async def get(self, evaluator_id: uuid.UUID) -> EvaluatorResponse:
        response = await self._client.request("GET", f"/v1/evaluators/{evaluator_id}")
        return EvaluatorResponse.model_validate(response.json())

    async def list(
        self, params: EvaluatorListParams | None = None
    ) -> Page[EvaluatorResponse]:
        params = params or EvaluatorListParams()
        response = await self._client.request(
            "GET",
            "/v1/evaluators",
            params=params.model_dump(mode="json", exclude_unset=True),
        )
        return Page[EvaluatorResponse].model_validate(response.json())

    async def iter(
        self, params: EvaluatorListParams | None = None
    ) -> AsyncIterator[EvaluatorResponse]:
        async for item in iterate_pages(params or EvaluatorListParams(), self.list):
            yield item

    async def update(
        self, evaluator_id: uuid.UUID, request: EvaluatorUpdateRequest
    ) -> EvaluatorResponse:
        response = await self._client.request(
            "PATCH",
            f"/v1/evaluators/{evaluator_id}",
            json=request.model_dump(mode="json", exclude_unset=True),
        )
        return EvaluatorResponse.model_validate(response.json())

    async def delete(self, evaluator_id: uuid.UUID) -> None:
        await self._client.request("DELETE", f"/v1/evaluators/{evaluator_id}")

    async def create_version(
        self, evaluator_id: uuid.UUID, request: EvaluatorVersionCreateRequest
    ) -> EvaluatorVersionResponse:
        response = await self._client.request(
            "POST",
            f"/v1/evaluators/{evaluator_id}/versions",
            json=request.model_dump(mode="json", exclude_unset=True),
        )
        return EvaluatorVersionResponse.model_validate(response.json())

    async def list_versions(
        self, evaluator_id: uuid.UUID, params: ListParams | None = None
    ) -> Page[EvaluatorVersionResponse]:
        params = params or ListParams()
        response = await self._client.request(
            "GET",
            f"/v1/evaluators/{evaluator_id}/versions",
            params=params.model_dump(mode="json", exclude_unset=True),
        )
        return Page[EvaluatorVersionResponse].model_validate(response.json())

    async def iter_versions(
        self, evaluator_id: uuid.UUID, params: ListParams | None = None
    ) -> AsyncIterator[EvaluatorVersionResponse]:
        async for item in iterate_pages(
            params or ListParams(),
            lambda page_params: self.list_versions(evaluator_id, page_params),
        ):
            yield item

    async def get_version(
        self, evaluator_id: uuid.UUID, version: int
    ) -> EvaluatorVersionResponse:
        response = await self._client.request(
            "GET", f"/v1/evaluators/{evaluator_id}/versions/{version}"
        )
        return EvaluatorVersionResponse.model_validate(response.json())

    async def update_version(
        self,
        evaluator_id: uuid.UUID,
        version: int,
        request: EvaluatorVersionUpdateRequest,
    ) -> EvaluatorVersionResponse:
        response = await self._client.request(
            "PATCH",
            f"/v1/evaluators/{evaluator_id}/versions/{version}",
            json=request.model_dump(mode="json", exclude_unset=True),
        )
        return EvaluatorVersionResponse.model_validate(response.json())

File: packages/core/src/resources/evaluators.ts
import { jsonBody, type KitaruTransport } from "../transport.js";
import type {
  EvaluatorCreateRequest,
  EvaluatorResponse,
  EvaluatorUpdateRequest,
  EvaluatorVersionCreateRequest,
  EvaluatorVersionResponse,
  EvaluatorVersionUpdateRequest,
  ListParams,
  Page,
  UnfilteredListParams,
} from "../types.js";
import {
  createPageValidator,
  validateEvaluator,
  validateEvaluatorVersion,
} from "./internal.js";
import {
  encodeListParams,
  iteratePages,
  type ResourceRequestOptions,
} from "./pagination.js";

const validateEvaluatorPage = createPageValidator(validateEvaluator);
const validateEvaluatorVersionPage = createPageValidator(
  validateEvaluatorVersion,
);

export class EvaluatorsResource {
  readonly #transport: KitaruTransport;

  constructor(transport: KitaruTransport) {
    this.#transport = transport;
  }

  async create(
    request: EvaluatorCreateRequest,
    options: ResourceRequestOptions = {},
  ): Promise<EvaluatorResponse> {
    return this.#transport.request({
      method: "POST",
      path: "/v1/evaluators",
      body: jsonBody(request),
      signal: options.signal,
      validate: validateEvaluator,
    });
  }

  async get(
    evaluatorId: string,
    options: ResourceRequestOptions = {},
  ): Promise<EvaluatorResponse> {
    return this.#transport.request({
      method: "GET",
      path: `/v1/evaluators/${encodeURIComponent(evaluatorId)}`,
      signal: options.signal,
      validate: validateEvaluator,
    });
  }

  async update(
    evaluatorId: string,
    request: EvaluatorUpdateRequest,
    options: ResourceRequestOptions = {},
  ): Promise<EvaluatorResponse> {
    return this.#transport.request({
      method: "PATCH",
      path: `/v1/evaluators/${encodeURIComponent(evaluatorId)}`,
      body: jsonBody(request),
      signal: options.signal,
      validate: validateEvaluator,
    });
  }

  async delete(
    evaluatorId: string,
    options: ResourceRequestOptions = {},
  ): Promise<void> {
    return this.#transport.request({
      method: "DELETE",
      path: `/v1/evaluators/${encodeURIComponent(evaluatorId)}`,
      responseType: "empty",
      signal: options.signal,
    });
  }

  async list(
    params: ListParams = {},
    options: ResourceRequestOptions = {},
  ): Promise<Page<EvaluatorResponse>> {
    return this.#transport.request({
      method: "GET",
      path: "/v1/evaluators",
      query: encodeListParams(params),
      signal: options.signal,
      validate: validateEvaluatorPage,
    });
  }

  iter(
    params: ListParams = {},
    options: ResourceRequestOptions = {},
  ): AsyncIterable<EvaluatorResponse> {
    return iteratePages(params, (pageParams) => this.list(pageParams, options));
  }

  async createVersion(
    evaluatorId: string,
    request: EvaluatorVersionCreateRequest,
    options: ResourceRequestOptions = {},
  ): Promise<EvaluatorVersionResponse> {
    return this.#transport.request({
      method: "POST",
      path: `/v1/evaluators/${encodeURIComponent(evaluatorId)}/versions`,
      body: jsonBody(request),
      signal: options.signal,
      validate: validateEvaluatorVersion,
    });
  }

  async getVersion(
    evaluatorId: string,
    version: number,
    options: ResourceRequestOptions = {},
  ): Promise<EvaluatorVersionResponse> {
    return this.#transport.request({
      method: "GET",
      path: `/v1/evaluators/${encodeURIComponent(evaluatorId)}/versions/${encodeURIComponent(String(version))}`,
      signal: options.signal,
      validate: validateEvaluatorVersion,
    });
  }

  async updateVersion(
    evaluatorId: string,
    version: number,
    request: EvaluatorVersionUpdateRequest,
    options: ResourceRequestOptions = {},
  ): Promise<EvaluatorVersionResponse> {
    return this.#transport.request({
      method: "PATCH",
      path: `/v1/evaluators/${encodeURIComponent(evaluatorId)}/versions/${encodeURIComponent(String(version))}`,
      body: jsonBody(request),
      signal: options.signal,
      validate: validateEvaluatorVersion,
    });
  }

  async listVersions(
    evaluatorId: string,
    params: UnfilteredListParams = {},
    options: ResourceRequestOptions = {},
  ): Promise<Page<EvaluatorVersionResponse>> {
    return this.#transport.request({
      method: "GET",
      path: `/v1/evaluators/${encodeURIComponent(evaluatorId)}/versions`,
      query: encodeListParams(params),
      signal: options.signal,
      validate: validateEvaluatorVersionPage,
    });
  }

  iterVersions(
    evaluatorId: string,
    params: UnfilteredListParams = {},
    options: ResourceRequestOptions = {},
  ): AsyncIterable<EvaluatorVersionResponse> {
    return iteratePages(params, (pageParams) =>
      this.listVersions(evaluatorId, pageParams, options),
    );
  }
}
